core/libp2p/connection/yamux/yamux_stream.cpp

```diff
@@ -0,0 +1,303 @@
+      YamuxedConnection::StreamId stream_id, uint32_t maximum_window_size)
+      : yamuxed_connection_{std::move(yamuxed_connection)},
+        stream_id_{stream_id},
+        maximum_window_size_{maximum_window_size} {}
+
+  void YamuxStream::read(gsl::span<uint8_t> out, size_t bytes,
+                         ReadCallbackFunc cb) {
+    if (bytes == 0 || out.empty() || static_cast<size_t>(out.size()) < bytes) {
+      return cb(Error::INVALID_ARGUMENT);
+    }
+    if (!is_readable_) {
+      return cb(Error::NOT_READABLE);
+    }
+    if (is_reading_) {
+      return cb(Error::IS_READING);
+    }
+
+    is_reading_ = true;
+
+    // if there is already enough data in our buffer, read it immediately and
+    // send an acknowledgement
+    if (read_buffer_.size() >= bytes) {
+      if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), bytes),
+                                   read_buffer_.data(), bytes)
+          != bytes) {
+        return cb(Error::INTERNAL_ERROR);
+      }
+      return yamuxed_connection_->streamAckBytes(
+          stream_id_, bytes,
+          [self{shared_from_this()}, cb = std::move(cb), bytes](auto &&res) {
+            self->is_reading_ = false;
+            if (!res) {
+              return cb(res.error());
+            }
+            self->read_buffer_.consume(bytes);
+            cb(bytes);
+          });
+    }
+
+    // else, set a callback, which is called each time a new data arrives
+    yamuxed_connection_->streamAddDataNotifyee(
+        stream_id_,
+        [self{shared_from_this()}, cb = std::move(cb), out, bytes]() mutable {
+          if (self->read_buffer_.size() < bytes) {
+            // not yet
+            return false;
+          }
+
+          if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), bytes),
+                                       self->read_buffer_.data(), bytes)
+              != bytes) {
+            cb(Error::INTERNAL_ERROR);
+          } else {
+            self->yamuxed_connection_->streamAckBytes(
+                self->stream_id_, bytes,
+                [self, cb = std::move(cb), out, bytes](auto &&res) {
+                  (void)out;
+                  self->is_reading_ = false;
+                  if (!res) {
+                    return cb(res.error());
+                  }
+                  self->read_buffer_.consume(bytes);
+                  cb(bytes);
+                });
+          }
+          return true;
+        });
+  }
+
+  void YamuxStream::readSome(gsl::span<uint8_t> out, size_t bytes,
+                             ReadCallbackFunc cb) {
+    if (bytes == 0 || out.empty() || static_cast<size_t>(out.size()) < bytes) {
+      return cb(Error::INVALID_ARGUMENT);
+    }
+    if (!is_readable_) {
+      return cb(Error::NOT_READABLE);
+    }
+    if (is_reading_) {
+      return cb(Error::IS_READING);
+    }
+
+    is_reading_ = true;
+
+    // if there is already enough data in our buffer, read it immediately
+    if (read_buffer_.size() != 0) {
```

Comment:
well, the only different thing is this condition. write a single method, which predicate "return condition", and apply it here.

e.g. generify and combine read/readSome methods

code duplication is root of all evil